File: CalcFraction/Java_Fraction/src/fraction/Fraction.java
package fraction;

public class Fraction {
    private int counter;
    private int denominator;

    public static void main(String[] args) {
        // TODO code application logic here
        System.out.println(new Fraction(Math.PI, 100));
        System.out.println(new Fraction(2, 6).add(new Fraction(2, 3)));
        System.out.println(new Fraction(2, 6).sub(new Fraction(3, 9)));
        System.out.println(new Fraction(2, 6).mul(new Fraction(3, 9)));
        System.out.println(new Fraction(2, 6).div(new Fraction(3, 9)));
    }

    public Fraction() {
        this(0, 1);
    }

    public Fraction(int counter) {
        this(counter, 1);
    }

    public Fraction(int counter, int denominator) {
        this.counter = counter;
        this.denominator = denominator;
        reduce();
    }

    public Fraction(double value, int maxdenom) {
        this.counter = (int) (value*maxdenom);
        this.denominator = maxdenom;
        reduce();
    }

    public Fraction add(Fraction other) {
        return new Fraction(this.counter*other.denominator + other.counter*this.denominator,
                this.denominator*other.denominator);
    }

    public Fraction sub(Fraction other) {
        return new Fraction(this.counter*other.denominator - other.counter*this.denominator,
                this.denominator*other.denominator);
    }

    public Fraction mul(Fraction other) {
        return new Fraction(this.counter*other.counter, this.denominator*other.denominator);
    }

    public Fraction div(Fraction other) {
        return new Fraction(this.counter*other.denominator, this.denominator*other.counter);
    }

    public double toDouble() {
        return 1.0 * this.counter / this.denominator;
    }

    public String toString() {
        return String.format("(%d, %d)", counter, denominator);
    }

    private void reduce() {
        int gcd = gcd(counter, denominator);
        if ( gcd != 0 ) {
            counter /= gcd;
            denominator /= gcd;
        }
        if ( denominator < 0 ) {
            counter *= -1;
            denominator *= -1;
        }
    }

    private static int gcd(int a, int b) {
        int h;
        if (a == 0) return Math.abs(b);
        if (b == 0) return Math.abs(a);

        do {
            h = a % b;
            a = b;
            b = h;
        } while (b != 0);

        return Math.abs(a);
    }
}

File: CalcFraction/Fraction.cpp
//Gruppe: Felix Steinke und Florian Grabowski

#include <istream>
#include <ostream>
#include <iostream>
#include "Fraction.h"


Fraction::Fraction(int counter, int denominator) {
    this->counter = counter;
    this->denominator = denominator;
    reduce();
}

Fraction::Fraction(double value, int maxdenom) {
    this->counter = (int) (value * maxdenom);
    this->denominator = maxdenom;
    reduce();
}

Fraction::~Fraction() {
    //The only time you have to worry about deallocating things is when you use new (C++) or malloc() (C).
}

Fraction Fraction::operator+(const Fraction &other) const {
    return Fraction(this->counter * other.denominator + other.counter * this->denominator,
                    this->denominator * other.denominator);
}

Fraction &Fraction::operator+=(const Fraction &other) {
    this->counter = this->counter * other.denominator + other.counter * this->denominator;
    this->denominator = this->denominator = this->denominator * other.denominator;
    reduce();
    return *this;
}

Fraction Fraction::operator-(const Fraction &other) const {
    return Fraction(this->counter * other.denominator - other.counter * this->denominator,
                    this->denominator * other.denominator);
}

Fraction &Fraction::operator-=(const Fraction &other) {
    this->counter = this->counter * other.denominator - other.counter * this->denominator;
    this->denominator = this->denominator * other.denominator;
    reduce();
    return *this;
}

Fraction Fraction::operator*(const Fraction &other) const {
    return Fraction(this->counter * other.counter,
                    this->denominator * other.denominator);
}

Fraction &Fraction::operator*=(const Fraction &other) {
    this->counter = this->counter * other.counter;
    this->denominator = this->denominator * other.denominator;
    reduce();
    return *this;
}

Fraction Fraction::operator/(const Fraction &other) const {
    return Fraction(this->counter * other.denominator,
                    this->denominator * other.counter);
}

Fraction &Fraction::operator/=(const Fraction &other) {
    int oldCounter = this->counter;
    int oldDenominator = this->denominator;
    int oldOtherCounter = other.counter;
    int oldOtherDenominator = other.denominator;
    this->counter = oldCounter * oldOtherDenominator;
    this->denominator = oldDenominator * oldOtherCounter;
    reduce();
    return *this;
}

bool Fraction::operator==(const Fraction &other) const {
    Fraction a = (double) *this;
    Fraction b = (double) other;
    return a.denominator == b.denominator && a.counter == b.counter;
}

Fraction::operator double() const {
    return 1.0 * this->counter / this->denominator;
}

void Fraction::reduce() {
    int gcd = Fraction::gcd(this->counter, this->denominator);
    if (gcd != 0) {
        counter /= gcd;
        denominator /= gcd;
    }
    if (denominator < 0) {
        counter *= -1;
        denominator *= -1;
    }
}

int Fraction::gcd(int a, int b) {
    int h;
    if (a == 0) return abs(b);
    if (b == 0) return abs(a);

    do {
        h = a % b;
        a = b;
        b = h;
    } while (b != 0);

    return abs(a);
}

std::ostream &operator<<(std::ostream &stream, const Fraction &fract) {
    return stream << '(' << fract.counter << "," << fract.denominator << ')';
}

std::istream &operator>>(std::istream &stream, Fraction &fract) {
    char ignore;
    return stream >> ignore >> fract.counter >> ignore >> fract.denominator >> ignore;
}